### DISCUSSION TITLE
Extract reusable testing package

### DISCUSSION THREAD

**Discussion by erickskrauch** [PHP-CS-Fixer/PHP-CS-Fixer#7694] [RFC]
Hello.

I'm the developer of the [custom rules](https://github.com/erickskrauch/php-cs-fixer-custom-fixers) for PHP-CS-Fixer. I want to maintain the high quality of those rules, so I write tests and run them the same way it is done in the main PHP-CS-Fixer repository. To do so I had to empirically figure out what dependencies are relatable to testing and which are not. I also had to request the PHP-CS-Fixer source code and enable autoloading for their `tests` directory. Take a look at the [diff](https://github.com/elyby/php-code-style/commit/18806e41e217a4c9d707d69ea134a182c38a6475#diff-d2ab9925cad7eac58e0ff4cc0d251a937ecf49e4b6bf57f8b95aab76648a9d34).

It took a while, but the deed was done and the tests started working.

I daresay there aren't many developers like me who will follow the same path. Besides, having gone down this path once, there is no guarantee that the PHP-CS-Fixer codebase won't change after a while and I won't have to search again for what dependencies are needed to make the tests work.

That's why I suggest to introduce a separate `php-cs-fixer/testing-tools` package, where all the support classes for testing fixers are placed (e.g. `AbstractFixerTestCase` and all related to it).

I'd be happy to help with shaping this package and then implementing it into the PHP-CS-Fixer codebase.

**Comment by Wirone** [comment]
Hi @erickskrauch! Interesting idea, however it needs to be decided by @keradus.

From my perspective I understand the need, but I don't know if a dedicated package for that is a good path. For example, [PHPStan provides a testing contract](https://github.com/phpstan/phpstan-src/tree/1.11.x/src/Testing) in its main src repo. Since it's in `src`, it's installed on the extensions' provider side when they require main PHPStan package. Then [it's used in tests](https://github.com/spaze/phpstan-disallowed-calls/blob/90fdf80dfd69fc7eebd3535e4a7b1aae273f1d00/tests/Calls/EchoCallsTest.php#L8).

For me such an approach is much better as you can use the same contract for the main tool and you don't need to update the Composer package and develop the testing API somewhere else.

What do you think?

**Reply by erickskrauch** [reply]
This was the case in the days of PHP-CS-Fixer v2. But with the release of v3, PHP-CS-Fixer started shipping with the testing module cut out.

Also, don't forget about the list of dependencies used inside `AbstractFixerTestCase`. You don't want to ship them as production dependencies, but if you go the route of keeping a single code base, the end developer will have to install about 6 dependencies themselves, which is one of the problems I originally described.

**Reply by Wirone** [reply]
Then I would rather see this extracted into local package (within the same repo), installed with [`path`](https://getcomposer.org/doc/05-repositories.md#path) type, which could be then released to separate repository on each Fixer version (with the same tag? under v1? I don't know at this point). Or at least it should start like this, and when we get stable testing API and see that it does not change often, we could move it to separate repo and tag as first stable contract for testing. I just wouldn't want to jump between repos to provide something in test base classes etc and to update Composer deps to see it in the Fixer 🙂.

**Reply by keradus** [reply]
reasoning for v2->v3 changes was to stop requiring PHPUnit as src dependency PHP CS Fixer, while also very, very limited usage of our testing utils.

Overall, I am all for provide official tooling for custom fixers testing, under BC contract - yet never managed to prioritise this myself.

idea I had was to:
- use monorepo approach and have dedicated package with explicit dependencies for test tooling, not to provide them as part of main package (And thus, require all testing deps like PHPUnit)
- install from local package, so we can have single PR changing both, tooling and usage
- use very same versioning (but skip version release if no content change for subpackage)

some early draft i created, probably worth nothing nowadays: https://github.com/keradus/PHP-CS-Fixer/pull/30/files#diff-0c394caae61462d3b67ff0fda3600ef26943d9bca640a0dedb5fb7d0ed639cf2R2

I was also wondering about usage of [subtreesplit.com](https://www.subtreesplit.com/) or [split.sh](https://github.com/splitsh/lite), to avoid any custom logic on our side for having this.

------

with limited need, I'm still not prioritising the idea on my side.

**Reply by Wirone** [reply]
Looks like we have very similar idea @keradus 🙂. I was also thinking about same-tag-but-only-if-changed policy, just like Symfony components have. Reasoning is exactly the same - to be able to provide changes in the testing contract _and_ actual Fixer changes within the same PR.

**Reply by keradus** [reply]
ah, one extra note - I already aimed to organise everything around what could be content of that "testing package" in single namespace - https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/tree/master/tests/Test

**Reply by aboks** [reply]
Is there any new progress on this, or an expected timeline? We would love to write tests for our custom fixers, so a testing-tools package that provides `AbstractFixerTestCase` would be most welcome.
